scipy/linalg/special_matrices.py

```diff
@@ -528,7 +528,7 @@ def block_diag(*arrs):
         raise ValueError("arguments in the following positions have dimension "
                          "greater than 2: %s" % bad_args)
 
-    shapes = np.array([a.shape for a in arrs])
+    shapes = np.array([a.shape if np.array(a.shape).all() else [0, 0] for a in arrs])
```

Comment:
Rather than iterating over the shape with `all`, it is probably going to be clearer if you do the check on `a.size`. I would also explicitly check for inequality to `0`, but that may just be me, i.e.:

```
shapes = np.array([a.shape if a.size > 0 else (0, 0) for a in arrs])
```